proposals/proposal_2020_12_rate-limiting.md

```diff
@@ -0,0 +1,199 @@
+
+2. Accepted: The proposal has been accepted and is ready for implementation.
+
+3. In Progress: An accepted proposal is being implemented by actual work.
+
+NOTE: The design is subject to change during this phase.
+
+4. Cancelled: While or before implementation the proposal was cancelled.
+
+NOTE: This can happen for a multitude of reasons.
+
+5. Complete: This feature/change is implemented.
+-->
+
+## Background
+
+<!-- Please leave this commented out section.
+
+This section is intended to describe the new feature, redesign or refactor.
+-->
+
+**Please provide a summary of the new feature, redesign or refactor:**
+
+This feature would attempt to limit the amount of running Vela builds at one time by a given party for system stability purposes.
+
+How is it today:
+- Vela customers have no way to achieve this.
+- Drone customers can set a `concurrency` value in their `.drone.yml` to limit deployments (_notably little use however_)
+- Jenkins customers can set a ton of options here, but are not competing for resources.
+
+
+**Please briefly answer the following questions:**
+
+1. Why is this required?
+
+- Protect system stability.
+  - "Noisy neighbors" - those who use a lot of Vela's resources with many commits/builds preventing other users from running builds.
+  - A `Restart build` storm - when a developer notices an issue with their build, often times the easiest troubleshooting step is to restart the build. Now imagine 100+ people doing this, but the issue was low system resources.
+- Allows customers to protect their deployments from being triggered twice. (a prod deploy for example)
+- Opens up future patterns that could be potentially harmful without these limits (scheduled builds for example)
+
+
+
+2. If this is a redesign or refactor, what issues exist in the current implementation?
+3. Are there any other workarounds, and if so, what are the drawbacks?
+4. Are there any related issues? Please provide them below if any exist.
+
+## Design
+
+**Please describe your solution to the proposal. This includes, but is not limited to:**
+
+* new/updated endpoints or url paths
+* new/updated configuration variables (environment, flags, files, etc.)
+* performance and user experience tradeoffs
+* security concerns or assumptions
+* examples or (pseudo) code snippets
+
+A few things to consider with this design (with options indented):
+- At what level should we limit?
+  - Org, total number for an org
+  - Repo, total number for a repo
+  - Pipeline, total number of pipelines with same org+repo+branch name (protecting deployments)
+- How could it be configurable?
+  - Admin set
+  - User set in .vela.yml
+  - User set in settings UI
+- What is the UX when the limit is reached?
+  - fail builds
+  - queue builds (can be broken down further for queue strategies)
+
+
+Given all above, we're left to pick and choose which parts of the above we like.
+
+
+#### Option MVP:
+```
+- System set max builds at the repo level.
```

Comment:
> Would the Vela administrator have the option to override this maximum value for specific repos?

Some of the use cases for this feature could be:
- a large team that works on a single repo which requires more builds to be able to function.
- a small team that is routinely at capacity for not valid reasons and should have less than default.

Is there any other use cases for this feature that you can think of?

Reply:
> a large team that works on a single repo which requires more builds to be able to function.

Only handling this use case within the MVP makes to me.

> a small team that is routinely at capacity for not valid reasons and should have less than default.

Always mixed opinions around "punishing" teams based on workflow but I do think this makes sense for a further iteration. Maybe having some sort of way within the UI to understand the current usage vs limit would help teams visualize the platform impact?

Reply:
> Always mixed opinions around "punishing" teams based on workflow but I do think this makes sense for a further iteration. Maybe having some sort of way within the UI to understand the current usage vs limit would help teams visualize the platform impact?

While I think it's outside the scope of this, I think having a metrics page that someone could go and look and see stats could be pretty interesting! I'll make a feature request for this to triage at a later date.

Reply:
> Would utilizing specific platforms and/or flavors create different maximum builds? Example: have lower maximum builds for MacOS agents since that hardware is limited compared to Linux agents.

I'm having a bit of a difficult time visualizing this. Would this be more to load balance between worker types?

Reply:
Maybe some example code would help visualize the thought:

```go
// define default build limits per agent type
build_limits := map[string]int{
		"windows": 5,
		"macos": 2,
		"linux": 10,
                "total": 10,
	}

// pseudo code to limit execution

if build_limits[payload["agent_type"] > len(<repo builds running agent_type>) {
  return fmt.Errorf("build limit for %s reached", payload["agent_type"]"
}
```

Basically allowing the Vela administrator to have further granularity over the limits for different types of agent. Arguably this would also be outside the scope of MVP but food for thought for future iterations.

Reply:
Ahhh that does make more sense. Thanks @JordanSussman. It kind of gives me an idea (that is also out of scope I feel):
Rate limiting exists in two modes:
- no limiting 
- rate limiting (whenever current builds is 90% of the total) according to what's already been defined above.

So that when Vela is underutilized people can run as many builds as they want, but if it's in high use it would protect itself.